src/async_wrap.cc

```diff
@@ -189,6 +191,18 @@ void AsyncWrap::EmitAfter(Environment* env, double async_id) {
        env->async_hooks_after_function());
 }
 
+static Maybe<double> GetAssignedPromiseAsyncId(Environment* env,
+                                               Local<Promise> promise,
+                                               Local<Value> id_symbol) {
+  Local<Value> maybe_async_id;
+  if (!promise->Get(env->context(), id_symbol).ToLocal(&maybe_async_id)) {
+    return Nothing<double>();
+  }
+  return maybe_async_id->IsNumber()
+      ? maybe_async_id->NumberValue(env->context())
+      : Nothing<double>();
```

Comment:
Sorry, I think I might have been a bit unclear: Empty `Maybe` and `MaybeLocal` values are used by the V8 API to propagate the information that an exception is pending. Our APIs should do the same – return `Nothing<>()` and empty `MaybeLocal<>()` only when an exception is pending, and a non-empty value otherwise (that can still indicate failure):

```suggestion
      : Just(AsyncWrap::kInvalidAsyncId);
```

If any call to a function that returns a `Maybe` or `MaybeLocal` returns an empty value, we should always return to JS as fast as possible, and not attempt any more operations that might call into JS – e.g. below, turn:

```c++
    Maybe<double> maybe_async_id = GetAssignedPromiseAsyncId(
        env, promise, env->async_id_symbol());
    Maybe<double> maybe_trigger_async_id = GetAssignedPromiseAsyncId(
        env, promise, env->trigger_async_id_symbol());
```

into

```c++
    double async_id;
    double trigger_async_id;
    if (!etAssignedPromiseAsyncId(env, promise, env->async_id_symbol()
            .ToLocal(&async_id))) return nullptr;
    if (!GetAssignedPromiseAsyncId(env, promise, env->trigger_async_id_symbol()
            .ToLocal(&trigger_async_id))) return nullptr;
```

I’ve tried to do a write-up of all this here (and if there’s room for improvements, that’s also great!): https://github.com/nodejs/node/blob/master/src/README.md#exception-handling

Reply:
My bad. 😓

Thanks for pointing me at the guide section. I'm going to update the PR.

Reply:
@addaleax could you take another look?